navigation/eskf/src/eskf.cpp

```diff
@@ -0,0 +1,210 @@
+                   v_n;
+
+    dR_dq.col(3) = 2 *
+                   ((e_3 * epsilon.transpose()) + (epsilon * e_3.transpose()) +
+                    (qw * skew(e_3))) *
+                   v_n;
+
+    Hx.block<3, 4>(0, 6) = dR_dq;
+
+    return Hx;
+}
+
+Eigen::Matrix3x18d ESKF::calculate_h_jacobian() {
+    Eigen::Matrix19x18d x_delta = Eigen::Matrix19x18d::Zero();
+    x_delta.block<6, 6>(0, 0) = Eigen::Matrix6d::Identity();
+    x_delta.block<4, 3>(6, 6) = calculate_q_delta();
+    x_delta.block<9, 9>(10, 9) = Eigen::Matrix9d::Identity();
+
+    Eigen::Matrix3x18d H = calculate_hx() * x_delta;
+    return H;
+}
+
+Eigen::Matrix3x1d ESKF::calculate_h() {
+    Eigen::Matrix3x1d h;
+    Eigen::Matrix3d R_bn =
+        current_nom_state_.quat.normalized().toRotationMatrix();
+
+    h = R_bn * current_nom_state_.vel;
+
+    return h;
+}
+
+void ESKF::nominal_state_discrete(const imu_measurement& imu_meas,
+                                  const double dt) {
+    Eigen::Vector3d acc =
+        current_nom_state_.quat.normalized().toRotationMatrix() *
+            (imu_meas.accel - current_nom_state_.accel_bias) +
+        current_nom_state_.gravity;
+    Eigen::Vector3d gyro = (imu_meas.gyro - current_nom_state_.gyro_bias) * dt;
+
+    current_nom_state_.pos = current_nom_state_.pos +
+                             current_nom_state_.vel * dt + 0.5 * sq(dt) * acc;
+    current_nom_state_.vel = current_nom_state_.vel + dt * acc;
+    current_nom_state_.quat =
+        (current_nom_state_.quat * vector3d_to_quaternion(gyro));
+    current_nom_state_.quat.normalize();
+    current_nom_state_.gyro_bias = current_nom_state_.gyro_bias;
+    current_nom_state_.accel_bias = current_nom_state_.accel_bias;
+    current_nom_state_.gravity = current_nom_state_.gravity;
+}
+
+void ESKF::error_state_prediction(const imu_measurement& imu_meas,
+                                  const double dt) {
+    Eigen::Matrix3d R = current_nom_state_.quat.normalized().toRotationMatrix();
+    Eigen::Vector3d acc = (imu_meas.accel - current_nom_state_.accel_bias);
+    Eigen::Vector3d gyro = (imu_meas.gyro - current_nom_state_.gyro_bias);
+
+    Eigen::Matrix18d A_c = Eigen::Matrix18d::Zero();
+    A_c.block<3, 3>(0, 3) = Eigen::Matrix3d::Identity();
+    A_c.block<3, 3>(3, 6) = -R * skew(acc);
+    A_c.block<3, 3>(6, 6) = -skew(gyro);
+    A_c.block<3, 3>(3, 9) = -R;
+    A_c.block<3, 3>(9, 9) = -Eigen::Matrix3d::Identity();
+    A_c.block<3, 3>(12, 12) = -Eigen::Matrix3d::Identity();
+    A_c.block<3, 3>(6, 12) = -Eigen::Matrix3d::Identity();
+    A_c.block<3, 3>(3, 15) = Eigen::Matrix3d::Identity();
+
+    Eigen::Matrix18x12d G_c = Eigen::Matrix18x12d::Zero();
+    G_c.block<3, 3>(3, 0) = -R;
+    G_c.block<3, 3>(6, 3) = -Eigen::Matrix3d::Identity();
+    G_c.block<3, 3>(9, 6) = Eigen::Matrix3d::Identity();
+    G_c.block<3, 3>(12, 9) = Eigen::Matrix3d::Identity();
+
+    auto [A_d, GQG_d] = van_loan_discretization(A_c, G_c, dt);
```

Comment:
This is a structured binding with Eigen:: types - auto and Eigen are not friends!!